### PR TITLE
[inductor] comprehensive padding

### PR DESCRIPTION
**Pull request by shunting314**
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* __->__ #120758

This PR adds the ability to pad tensor strides during lowering. The goal is to make sure (if possible) tensors with bad shape can have aligned strides so GPU can access the memory more efficiently.

By testing BlenderbotSmallForConditionalGeneration I already see 2.5ms speedup.

cc @voznesenskym @penguinwu @EikanWang @jgong5 @Guobing-Chen @XiaobingSuper @zhuhaozhe @blzheng @wenzhe-nrv @jiayisunx @peterbell10 @ipiszy @yf225 @chenyang78 @kadeng @muchulee8 @aakhundov @ColinPeppler @amjames @desertfire @chauhang @jansel @Chillee @eellison

### CODE REVIEW

**Review by eellison — COMMENTED** [review]
Cool!

- We might need some logic around not increasing memory too much. Not sure if this would actually be an issue in practice.
- I think we still want to preserve strides for user visible outputs.
- We should make sure that we're still appropriately passing in dense tensors to extern ops that require it.. I think this will just work but worth double checking. 
- Not sure if there is an aot autograd related issue of changing strides... I dont think there is but I remember it has special logic to deal with changing strides from inductor

**Comment by shunting314** [comment]
Current result [link](https://hud.pytorch.org/benchmark/compilers?startTime=Tue%2C%2012%20Mar%202024%2017%3A40%3A05%20GMT&stopTime=Tue%2C%2019%20Mar%202024%2017%3A40%3A05%20GMT&granularity=hour&suite=torchbench&mode=training&dtype=amp&lBranch=gh/shunting314/103/head&lCommit=5614f05f8ebf8115e94d1e7acfb58649add0daf9&rBranch=main&rCommit=65ccac6f175912062926886ebc86ee6b09a7ce23)
- HF 4% speedup
- TIMM/TB 4% slowdown

I debugged the slowdown for nvidia_deeprecommender. The root cause is we change the stride of variable/gradients which can result in an extra copy in AccumulateGrad. The point of the copy is to make sure variable and its gradient have the same layout so optimizer can be more efficient. But with tiling, a compiled optimizer does not really benefit a lot from that. I added an option to skip the extra copy in AccumulateGrad and bring back the perf for nvidia_deeprecommender. Will try other slowed models. Perf for nvidia_deeprecommender:
- no padding: 10.9ms
- with padding: 11.68ms
- with padding and skip gradient layout contract: 10.63ms. It brings back the perf and shows some speedup.

**Comment by Chillee** [comment]
I'm pretty skeptical about adding a mode to autograd that relaxes this restriction 🤔 Why can't you just disable the layout optimization for gradient computation?

**Comment by shunting314** [comment]
> I'm pretty skeptical about adding a mode to autograd that relaxes this restriction

What's the concern here? From the code comment, the grad layout contract is mainly to make the optimizer (and c10d reducer as mentioned by the comment. But I'm not sure if that's very relevant here) efficient. But with an compiled optimizer, let variable and gradient having different layout is not a big problem.

> Why can't you just disable the layout optimization for gradient computation?

I think that's possible. I'm just worried that if we restrict too much on what tensors we can change the layout, we would get less speedup. But if we think keeping gradients' layout the same as eager is important, we can do that. On the other hand, I need a way to figure out what tensors in the bwd graph are gradients.

**Comment by Chillee** [comment]
> On the other hand, I need a way to figure out what tensors in the bwd graph are gradients.

All the outputs of the backwards graph are gradients.

**Comment by albanD** [comment]
Chiming in here quickly. Disabling the grad layout is indeed a bit problematic.
In general, we aim to preserve layout of the objects going back to the user in compile. And for AccumulateGrad that has a very complex semantic, we went through quite a bit of trouble to use the original function instead of trying to duplicate to guarantee exact 1-1 correspondence between the two.
This flag would definitely defeat that and if we're going down this path of breaking the layout preservation, we should just have accumulate grad done directly within compile and not re-use accumulate grad at all instead of adding a flag there.

That being said, I don't think you actually want to break that contract?
One option depending on your use case is that if the .grad field is already allocated, we will re-use at as-is though. So you can set the .grad field ahead of time (or zero it out instead of setting it to None between allocations) to preserve a given layout there.

**Comment by shunting314** [comment]
Screenshot 2024-04-05 at 10 35 33 AM

Perf with cudagraphs looks good now. The perf without cudagraps for TB still has large regress though

**Comment by shunting314** [comment]
Screenshot 2024-04-06 at 10 21 08 PM

All perf regress has been fixed now. It turns out that the large regress from default setting from TB is due to using a new triton pin. It seems that triton regresses.. Reverting the pin change eventually fixes it. But this also means we are not using Hongtao's fix here: https://github.com/openai/triton/pull/3497 (since I need use a relatively new triton to apply the PR). I'll try to dig more.

On the other hand, I'll start cleaning up this PR to make it ready to review.

**Comment by shunting314** [comment]
@jansel , @Chillee , @eellison  just a heads up that this PR is ready for review.

File: torch/_inductor/compile_fx.py [hunk 1]
```diff
@@ -322,6 +322,22 @@ def is_tf32_warning_applicable(gm: torch.fx.GraphModule):
...
+    """
+    For CPU backend, enable comprehensive padding causes some unit tests
+    fail due to changing number of generated kernels. Skip for now.
+    """
+    has_cuda = any(
+        t.device.type == "cuda" for t in example_inputs if isinstance(t, torch.Tensor)
+    )
+
+    if config.comprehensive_padding and not has_cuda:
+        perf_hint_log.warning("Skip comprehensive padding on CPU")
```
Comment (jansel) [inline torch/_inductor/compile_fx.py:335]: This doesn't seem like it should be a warning.

File: torch/_inductor/compile_fx.py [hunk 2]
```diff
@@ -1344,6 +1364,15 @@ def partition_fn(graph, joint_inputs, **kwargs):
...
     def bw_compiler(model: torch.fx.GraphModule, example_inputs: List[torch.Tensor]):
+        user_visible_outputs = set()
+
+        if config.bw_outputs_user_visible:
+            *_, model_outputs_node = model.graph.nodes
+            assert model_outputs_node.op == "output"
+            model_outputs = pytree.arg_tree_leaves(*model_outputs_node.args)
+            user_visible_outputs = {
+                n.name for n in model_outputs if isinstance(n, torch.fx.Node)
+            }
```
Comment (jansel) [inline torch/_inductor/compile_fx.py:1375]: sets have non-determinstic order, which can lead to annoying issues. Suggest `dict.fromkeys`

File: torch/_inductor/ir.py [hunk 3]
```diff
@@ -2573,7 +2592,94 @@ def is_channels_last_stride_ordered(self):
...
+    def _pad_strides(in_strides, size, align=DEFAULT_ALIGN):
+        """
+        The padding does not change stride order but makes sure all strides larger
+        than the threshold are multiple of align.
+        """
+        if len(in_strides) == 0:
+            return in_strides
+
+        # a loose check for channels last
+        if not config.pad_channels_last and len(in_strides) == 4 and in_strides[1] == 1:
```
Comment (jansel) [inline torch/_inductor/ir.py:2605]: I think there is a helper elsewhere in this file for this check, reuse or refactor it?

File: torch/_inductor/metrics.py [hunk 4]
```diff
@@ -43,6 +43,8 @@
 # Each element counts the number of inner kernels in each outer loop fusion.
 cpp_outer_loop_fused_inner_counts: List[int] = []
 
+num_comprehensive_padding = 0
```
Comment (jansel) [inline torch/_inductor/metrics.py:46]: add to `reset()`,etc below

**Review by jansel — CHANGES_REQUESTED** [review]

File: torch/_inductor/compile_fx.py [hunk 5]
```diff
@@ -336,7 +352,9 @@ def count_bytes_inner(
         _recursive_post_grad_passes(gm, False)
 
     graph = GraphLowering(gm, shape_env=shape_env, num_static_inputs=num_fixed)
-    with V.set_graph_handler(graph), V.set_real_inputs(example_inputs):
+    with V.set_graph_handler(graph), V.set_real_inputs(
+        example_inputs
+    ), maybe_disable_comprehensive_padding(example_inputs):
```
Comment (eellison) [inline torch/_inductor/compile_fx.py:357]: any reason to do this as a config patch instead of as a field on GraphLowering?

File: torch/_inductor/compile_fx.py [hunk 6]
```diff
@@ -1344,6 +1364,15 @@ def partition_fn(graph, joint_inputs, **kwargs):
     @dynamo_utils.dynamo_timed
     @dynamo_utils.maybe_cprofile
     def bw_compiler(model: torch.fx.GraphModule, example_inputs: List[torch.Tensor]):
+        user_visible_outputs = set()
+
+        if config.bw_outputs_user_visible:
+            *_, model_outputs_node = model.graph.nodes
```
Comment (eellison) [inline torch/_inductor/compile_fx.py:1370]: nit: `torch._inductor.utils.output_node`

File: torch/_inductor/config.py [hunk 7]
```diff
@@ -421,6 +421,16 @@ def decide_compile_threads():
...
 
+# Control if we will do padding for pointwise/reductions
+comprehensive_padding = (
+    os.environ.get("TORCHINDUCTOR_COMPREHENSIVE_PADDING", "1") == "1"
+)
+pad_channels_last = False
+
+# Whether to treat output of the backward graph as user visible.
+# For user visible outputs, inductor will make sure the stride matches with eager.
+bw_outputs_user_visible = True
```
Comment (eellison) [inline torch/_inductor/config.py:432]: Can you say more why you needed to do this ? I saw alban comment something along the lines of gradient stealing or something?

File: torch/_inductor/graph.py [hunk 8]
```diff
@@ -148,6 +149,61 @@ def getattr_recursive(obj, target):
...
+    going thru a reduction.
+    """
+    if not config.comprehensive_padding:
+        return
+    ops_dislike_padding = {
+        aten.convolution,
+        aten.convolution_backward,
+    }
+    # what's a better way to collect the reduction ops?
+    ops_like_padding = {
```
Comment (eellison) [inline torch/_inductor/graph.py:168]: It would be great (not needed for this PR per se) to have a `reduction` tag similar to how we have a `pointwise` tag. This was also recently a need in @Chillee's aot autograd pr.

File: torch/_inductor/graph.py [hunk 9]
```diff
@@ -148,6 +149,61 @@ def getattr_recursive(obj, target):
     return attr_itr
 
 
+def mark_nodes_dislike_padding(g):
+    """
+    Nodes like convolution/convolution_backward want its input to be contiguous.
```
Comment (eellison) [inline torch/_inductor/graph.py:154]: nit: s/contiguous/dense

File: torch/_inductor/ir.py [hunk 10]
```diff
@@ -2573,7 +2592,94 @@ def is_channels_last_stride_ordered(self):
...
+    @staticmethod
+    def _pad_strides(in_strides, size, align=DEFAULT_ALIGN):
+        """
+        The padding does not change stride order but makes sure all strides larger
+        than the threshold are multiple of align.
+        """
+        if len(in_strides) == 0:
+            return in_strides
+
+        # a loose check for channels last
```
Comment (eellison) [inline torch/_inductor/ir.py:2604]: Should we check whether or not `layout_opt` is True to distinguish between whether or not we are considering `size[1] == 1` as channels last or not ?

File: torch/_inductor/ir.py [hunk 11]
```diff
@@ -2573,7 +2592,94 @@ def is_channels_last_stride_ordered(self):
...
+
+        current_fx_node = V.get_current_node()
+        if hasattr(current_fx_node, "meta") and current_fx_node.meta.get(
+            "dislike_padding", False
+        ):
+            return in_strides
+
+        # get_stride_order does not work with dynamic shape. Also we can not
+        # statically decide if a padding is needed or how much padding we should
+        # do for dynamic shape.
```
Comment (eellison) [inline torch/_inductor/ir.py:2616]: Couldn't we pad to an aligned multiple even with dynamic shape ?

File: torch/_inductor/ir.py [hunk 12]
```diff
@@ -2573,7 +2592,94 @@ def is_channels_last_stride_ordered(self):
...
+            isinstance(s, (int, sympy.Integer))
+            for s in itertools.chain(in_strides, size)
+        ):
+            return in_strides
+
+        stride_order = get_stride_order(in_strides)
+        fill_order = stride_order2fill_order(stride_order)
+
+        new_strides = [0 for _ in range(len(in_strides))]
+        # since way pad when the layout is flexible, we can decide the
```
Comment (eellison) [inline torch/_inductor/ir.py:2629]: nit: /s/way/we

File: torch/_inductor/ir.py [hunk 13]
```diff
@@ -2573,7 +2592,94 @@ def is_channels_last_stride_ordered(self):
...
+        # smallest stride to be 1.
+        new_strides[fill_order[0]] = 1
+
+        # don't align an too small stride since that cause too much memory increase.
```
Comment (eellison) [inline torch/_inductor/ir.py:2633]: nit: don't align a too small

File: torch/_inductor/ir.py [hunk 14]
```diff
@@ -2573,7 +2592,94 @@ def is_channels_last_stride_ordered(self):
...
+        # This change turnes HF AllenaiLongformerBase amp training from a loss of 1.09x to a win of 1.05x.
+        # (baseline: 71.09ms, padding w/o this change: 77.38ms, padding with this change: 67.77ms)
+        align_stride_threshold = 1024
+        padded = False
+        for rank, idx in enumerate(fill_order[1:], start=1):
+            prev_idx = fill_order[rank - 1]
+            stride = new_strides[prev_idx] * size[prev_idx]
+
+            if stride > align_stride_threshold and stride % align != 0:
+                stride = (stride + align - 1) // align * align
```
Comment (eellison) [inline torch/_inductor/ir.py:2656]: Nit: ceildiv(stride, align) * align

File: torch/_inductor/config.py (hunk 7 above)
Comment (shunting314) [reply torch/_inductor/config.py:432]: Previously inductor does not guarantee that the output of backward graph (e.g. gradients for parameters or graph inputs if they require gradients) have the same strides as eager. This may ends up with the case that a gradient has a different layout compared to the corresponding parameter. The PyTorch autograd engine has something called gradient layout contract that may do a extra copy in that case so that the strides of gradient and parameter match. This makes the optimizer more efficient. Treating output of backward graph user visible can avoid that extra copy.

On the other hand, it's just good to make the strides of gradients matching eager in case user runs some view op on them.

File: torch/_inductor/graph.py (hunk 9 above)
Comment (shunting314) [reply torch/_inductor/graph.py:154]: Hmm, what would be the difference between 'dense' and 'contiguous'? I see convolution/convolution_backward has calls like `tensor.contiguous(memory_format=the_format)`

File: torch/_inductor/ir.py (hunk 11 above)
Comment (shunting314) [reply torch/_inductor/ir.py:2616]: We are not simply padding each stride here. We need consider stride order (check test_pad_3d_tensor for why that's the case). For dynamic shape, get_stride_order just does not work.

There may be other complexity that I have not considered yet.

Also I found we don't pad mm for dynamic shape either :)

File: torch/_inductor/compile_fx.py (hunk 5 above)
Comment (shunting314) [reply torch/_inductor/compile_fx.py:357]: Nope, either way should work. I can access the config by either config.comprehensive_padding or V.graph.comprehensive_padding .

File: torch/_inductor/graph.py (hunk 9 above)
Comment (eellison) [reply torch/_inductor/graph.py:154]: Sometimes contiguous is taken to assume `torch.contiguous(memory_format=torch.contiguous_format)`. Whereas dense is more clear it might be channels last or contiguous. It doesn't really matter, either is fine.

File: torch/_inductor/ir.py (hunk 10 above)
Comment (shunting314) [reply torch/_inductor/ir.py:2604]: I just refactored this code to call `Layout.is_channels_last_contiguous` as Jason suggested above. We now check for channels last more accurately.

I think this check makes sense no matter if `layout_opt` is on or off.

File: torch/_inductor/graph.py (hunk 9 above)
Comment (shunting314) [reply torch/_inductor/graph.py:154]: Got it. Make sense to me!

**Comment by shunting314** [comment]
> All perf regress has been fixed now. It turns out that the large regress from default setting from TB is due to using a new triton pin. It seems that triton regresses.. Reverting the pin change eventually fixes it. But this also means we are not using Hongtao's fix here: https://github.com/openai/triton/pull/3497 (since I need use a relatively new triton to apply the PR). I'll try to dig more.

Since Hongtao's fix may take a while to get in (still under review). And our triton pin also needs some time to be updated (even if we have all the fixes in right now, we may still need run meta internal tests. That takes time). I'll try if I can apply hongtao's fix  in the current triton pin inductor uses and measure how that works together with comprehensive padding.

**Comment by eellison** [comment]
As a follow up would be great to see if we could use this to replace some of the current pad_mm logic.

**Comment by shunting314** [comment]
> As a follow up would be great to see if we could use this to replace some of the current pad_mm logic.

I think it depends on a few things
1. we can not pad weights since they have FIxedLayout. 
2. for activations, I think this PR can help pad the strides. But we'll need benchmark if padded tensor has as good matmul performance as a tensor with aligned shape.

**Comment by shunting314** [comment]
cc @jansel  to take another look

**Review by jansel — APPROVED** [review]
